From fbdc154a59a4fa843f4742a901ea6a20e5ccaac0 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 30 Oct 2024 19:42:54 +0100
Subject: [PATCH] usb: dwc2: fix shutdown in peripheral mode when gadget isn't
 configured

When in peripheral mode, the low level HW may have been left disabled
from the probe, e.g.:
	/* Gadget code manages lowlevel hw on its own */
	if (hsotg->dr_mode == USB_DR_MODE_PERIPHERAL)
		dwc2_lowlevel_hw_disable(hsotg);

It can be the case when no gadget configuration has been applied.
It could happen if the gadget has been removed before suspending.

When going to shutdown, register access in dwc2_is_device_mode(dwc2) will
trigger an exception:

E/TC:0   stm32_serc_handle_ilac:133 SERC exceptions [191:160]: 0x2000000
E/TC:0   stm32_serc_handle_ilac:139 SERC exception ID: 185
[   99.262259] Internal error: synchronous external abort: 0000000096000210 [#1] PREEMPT SMP
[   99.264829] Modules linked in:
[   99.267860] CPU: 0 PID: 468 Comm: reboot Not tainted 6.6.48 #7
[   99.273701] Hardware name: STMicroelectronics STM32MP215F-DK Discovery Board (DT)
[   99.281146] pstate: 60000005 (nZCv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   99.288093] pc : dwc2_disable_global_interrupts+0x8/0x50
[   99.293446] lr : dwc2_driver_shutdown+0x1c/0x34
[   99.297883] sp : ffff8000822ebc10
[   99.301205] x29: ffff8000822ebc10 x28: ffff000007146740 x27: 0000000000000000
[   99.308361] x26: ffff8000813d1d58 x25: 0000000000000001 x24: ffff800081aa6028
[   99.315414] x23: ffff000004290890 x22: ffff800081b0d7a0 x21: ffff00000428bc10
[   99.322568] x20: ffff000004290810 x19: ffff0000043b6080 x18: ffffffffffffffff
[   99.329622] x17: 0000000000000000 x16: 0000000000000000 x15: 00001e792d2d8ca8
[   99.336776] x14: 00001e792d2d8ca8 x13: 0000000000000094 x12: ffff000005f6d11c
[   99.343929] x11: 0000000000000040 x10: ffff80008190d658 x9 : ffff80008190d650
[   99.350983] x8 : ffff000004400028 x7 : 0000000000000000 x6 : 0000000000000000
[   99.358136] x5 : ffff000004290848 x4 : 0000000000000000 x3 : 0000000000000002
[   99.365189] x2 : 0000000000000001 x1 : ffff800082090008 x0 : ffff0000043b6080
[   99.372344] Call trace:
[   99.374764]  dwc2_disable_global_interrupts+0x8/0x50
[   99.379704]  platform_shutdown+0x24/0x34
[   99.383638]  device_shutdown+0x150/0x25c
[   99.387472]  kernel_restart+0x40/0xc0
[   99.391109]  __do_sys_reboot+0x178/0x268
[   99.395044]  __arm64_sys_reboot+0x24/0x30
[   99.399079]  invoke_syscall+0x48/0x114
[   99.402814]  el0_svc_common.constprop.0+0x40/0xe0
[   99.407452]  do_el0_svc+0x1c/0x28
[   99.410783]  el0_svc+0x40/0xe4
[   99.413816]  el0t_64_sync_handler+0x120/0x12c
[   99.418149]  el0t_64_sync+0x190/0x194
[   99.421787] Code: b9000001 d65f03c0 f9400401 91002021 (b9400021)
[   99.427826] ---[ end trace 0000000000000000 ]---

It is due to all clocks being disabled. In such case, clocks are enabled
temporarily before irq shutdown

Fixes: c846b03ff767 ("USB: dwc2: Don't turn off the usbphy in suspend if wakeup is enabled")
Change-Id: If9b470729e3b03583db3006ef2467ed475bd81ff
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Signed-off-by: Ram Dayal <ram.dayal@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/419009
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/usb/dwc2/platform.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -379,8 +379,18 @@ static void dwc2_driver_shutdown(struct
 {
 	struct dwc2_hsotg *hsotg = platform_get_drvdata(dev);
 
+	/*
+	 * The low level HW may have been left disabled from the probe,
+	 * or disabled later on, enable it before turning off interrupts
+	 */
+	if (!hsotg->ll_hw_enabled)
+		__dwc2_lowlevel_hw_enable(hsotg);
+
 	dwc2_disable_global_interrupts(hsotg);
 	synchronize_irq(hsotg->irq);
+
+	if (!hsotg->ll_hw_enabled)
+		__dwc2_lowlevel_hw_disable(hsotg);
 }
 
 /**
